File: DSA/DSA01028 LIỆT KÊ TỔ HỢP.cpp
#include <bits/stdc++.h>
using namespace std;

long long c(int n, int k)
{
    int total = 1;
    for (int i = 1; i <= k; i++)
    {
        total *= n-i+1;
        total /= i;
    }
    return total;
}

void update(vector<int> &A, int n, int k)
{
    int i = k-1;
    while (A[i] == n-k+i+1)
    {
        i--;
    }
    if (i < 0)
    {
        return;
    }
    
    A[i]++;
    for (int j = i+1; j < k; j++)
    {
        A[j] = A[i]+j-i;
    }
    
}

void submit()
{
    int n, k;
    cin >> n >> k;
    set<int> A;
    for (int i = 0; i < n; i++)
    {
        int temp;
        cin >> temp;
        A.insert(temp);
    }
    vector<int> B(A.begin(), A.end());
    vector<int> C(k);

    for (int i = 0; i < k; i++)
    {
        C[i] = i+1;
    }
    
    n = A.size();
    int total = c(n, k);

    for (int i = 0; i < total; i++)
    {
        for (auto &&i : C)
        {
            cout << B[i-1] << " ";
        }
        cout << endl;
        update(C, n, k);
    }
    
}   

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    submit();

    return 0;
}

File: DSA_python/DSA01028 LIỆT KÊ TỔ HỢP.py


def c(n, k):
    total = 1
    for i in range(1, k+1):
        total *= n-i+1
        total /= i
    return int(total)

def update(A, n, k):
    i = k-1
    while A[i] == n-k+i+1:
        i -= 1
    
    if i < 0:
        return
    
    A[i] += 1
    for j in range(i+1, k):
        A[j] = A[i]+j-i

def submit():
    n, k = map(int, input().split())
    A = set(map(int, input().split()))

    B = []
    for i in A:
        B.append(i)
    B.sort()
    C = [i for i in range(1, k+1)]
    n = len(B)
    
    for i in range(c(n, k)):
        for i in C:
            print(B[i-1], end=" ")
        print()
        update(C, n, k)

submit()